src/main/kotlin/org/elm/ide/refactoring/nameSuggestions.kt

```diff
@@ -0,0 +1,70 @@
+package org.elm.ide.refactoring
+
+import com.intellij.psi.codeStyle.NameUtil
+import org.elm.lang.core.psi.ElmExpressionTag
+import org.elm.lang.core.psi.elements.ElmFieldAccessExpr
+import org.elm.lang.core.psi.elements.ElmFunctionCallExpr
+import org.elm.lang.core.psi.elements.ElmValueExpr
+import org.elm.lang.core.resolve.scope.ExpressionScope
+import org.elm.lang.core.toElmLowerId
+import org.elm.lang.core.types.TyFloat
+import org.elm.lang.core.types.TyFunction
+import org.elm.lang.core.types.TyInt
+import org.elm.lang.core.types.findTy
+
+data class SuggestedNames(val default: String, val all: LinkedHashSet<String>)
+
+/**
+ * Return at least one suggestion for a name that describes the receiver expression.
+ */
+fun ElmExpressionTag.suggestedNames(): SuggestedNames {
+    val names = LinkedHashSet<String>()
+    val ty = findTy()
+
+    when {
+        this is ElmFunctionCallExpr -> {
+            // suggest based on function call (e.g. suggest "books" for expr: `getBooks "shakespeare" library`)
+            val target = target
+            if (target is ElmValueExpr) {
+                names.addName(target.referenceName)
+            }
+        }
+        this is ElmFieldAccessExpr -> {
+            // suggest the last field in a record field access chain (e.g. "title" in expr `model.currentPage.title`)
+            names.addName(this.lowerCaseIdentifier?.text)
+        }
+    }
+
+    // if present, suggest type alias (useful for records which often have a descriptive alias name)
+    ty?.alias?.name?.let { names.addName(it) }
+
+    // suggest based on common types
+    when (ty) {
```

Comment:
Thanks. I did everything except special-case `TyFunction`. I don't see what you're getting at by checking the return type. Are you assuming that the user is going to apply the args within the function body? My assumption is that they would instead add parameters to the function declaration in the `let` (it would be nice to do that automatically, but not now).

Reply:
My thought was that if you extract, say, a partial application of your `Model` constructor, it would make sense to call the function `model`, since that's the value of the extracted function after you add parameters. Regardless, people probably don't use the default names often, so it doesn't matter either way.